url5583f3dbb2_XOR19_XScript_tgz-pJ8-xscript_interactive_MainJ8/src/xscript/XExec.java

```diff
@@ -0,0 +1,1927 @@
+		} else if (XUtils.isInstanceOf(rt, m, rt.getBaseType(XUtils.TYPE))) {
+			params.add(0, m);
+			callHelper2("___new__", params, map);
+		} else {
+			if (handles__getItem__(m)) {
+				callHelper2("___call__", params, map);
+			} else {
+				XValue newMethod = XUtils.lookup(rt, m, "__call__", XValue.REF_NONE);
+				call(newMethod, m, params, map);
+			}
+		}
+	}
+
+	private void call(XValue m, XValue thiz, XValue[] args, boolean visibleToStacktrace) {
+		callFrame = new XCallFrame(rt, callFrame, m, stackpointer, visibleToStacktrace);
+		push(thiz);
+		for (XValue arg : args) {
+			push(arg);
+		}
+	}
+
+	public boolean updateWaiting() {
+		if (timeout > 0) {
+			long time = rt.getTime();
+			timeout += lastTime - time;
+			lastTime = time;
+			if (timeout <= 0) {
+				timeout = 0;
+				push(XValueInt.valueOf(1));
+			}
+		}
+		return timeout == 0;
+	}
+
+	public int run(int instrs) {
+		if (instrs == 0)
+			return 0;
+		while ((instrs == -1 || instrs > 0) && timeout == 0 && rt.isRunning()) {
+			if (instrs > 0)
+				instrs--;
+			while (callFrame.isFinished()) {
+				if (stackpointer == callFrame.basepointer) {
+					push(XValueNull.NULL);
+				} else if (stackpointer > callFrame.basepointer + 1) {
+					XValue ret = pop();
+					pop(stackpointer - callFrame.basepointer);
+					push(ret);
+				}
+				callFrame = callFrame.getParent();
+				if (callFrame == null)
+					return instrs;
+			}
+			if (exception != null) {
+				XCatchHandler catchHandler = callFrame.getCatchHandler();
+				while (catchHandler == null) {
+					callFrame = callFrame.getParent();
+					if (callFrame == null)
+						return instrs;
+					catchHandler = callFrame.getCatchHandler();
+				}
+				int stackpointer = catchHandler.getStackPointer();
+				pop(this.stackpointer - stackpointer);
+				push(exception);
+				exception = null;
+				callFrame.jumpTo(catchHandler.getInstructionPointer());
+			}
+			try {
+				exec();
+			} catch (XRuntimeException e) {
+				XValue value = rt.getModule("sys");
+				XValue type = value.getRaw(rt, e.getType());
+				XValue exc = rt.alloc(type);
+				XValue message = rt.alloc(e.getMessage());
+				XValue stackTrace = XUtils.convertStackTrace(rt, getStackTrace(false));
+				exc.setRaw(rt, "_message", message);
+				exc.setRaw(rt, "_cause", XValueNull.NULL);
+				exc.setRaw(rt, "_stackTrace", stackTrace);
+				exception = exc;
+			} catch (Throwable e) {
+				StringWriter sw = new StringWriter();
+				e.printStackTrace(new PrintWriter(sw));
+				String message = sw.toString();
+				XValue value = rt.getModule("sys");
+				XValue type = value.getRaw(rt, "NativeError");
+				XValue exc = rt.alloc(type);
+				XValue m = rt.alloc(message);
+				XValue stackTrace = XUtils.convertStackTrace(rt, getStackTrace(false));
+				exc.setRaw(rt, "_message", m);
+				exc.setRaw(rt, "_cause", XValueNull.NULL);
+				exc.setRaw(rt, "_stackTrace", stackTrace);
+				exception = exc;
+			}
+		}
+		return instrs;
+	}
+
+	private void exec() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 220. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.